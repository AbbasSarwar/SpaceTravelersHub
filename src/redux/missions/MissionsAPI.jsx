import { createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
export const missionsData = createAsyncThunk('get/fetchmissions', async () => {
    try {
      const response = await axios.get('https://api.spacexdata.com/v3/missions');
      const data = response.data
      return data.map((mission)=>({
        id: mission.mission_id,
        mission_name: mission.mission_name,
        description: mission.description,
    }));
    } catch (error) {
      throw Error('Error fetching data from API');
    }
  });