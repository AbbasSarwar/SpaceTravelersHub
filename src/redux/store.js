import { configureStore } from "@reduxjs/toolkit";
import { Rocket } from "./rockets/RocketsSlice";
import missionsReducer from './missions/missionsSlice';
export const store = configureStore({
    reducer: {
        rockets: Rocket.reducer,
        missions: missionsReducer,
    },
});
export default store;

