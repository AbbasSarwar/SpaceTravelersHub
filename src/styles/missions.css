table,
th,
td {
  border: 1px solid black;
  border-collapse: collapse;
}

table {
  width: 100%;
  margin-top: 2rem;
}

th,
td {
  padding: 1rem;
}

td {
  text-align: center;
}

tbody tr:nth-child(odd) {
  background-color: rgb(231, 231, 231);
}

.mission {
  width: 10%;
}

.description {
  width: 60%;
}

.status {
  width: 12%;
}

.join-leave-mission {
  width: 12%;
}

.join-btn {
  color: black;
  cursor: pointer;
  padding: 0.5rem;
}

.status-badge {
  background-color: rgb(97, 97, 97);
  color: white;
  padding: 0.3rem;
  font-size: 0.8rem;
}
