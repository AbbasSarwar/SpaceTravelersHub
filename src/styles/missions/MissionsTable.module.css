.join {
  color: black;
  cursor: pointer;
  padding: 0.5rem;
}

.leave {
  color: black;
  cursor: pointer;
  padding: 0.5rem;
  border: 1px solid red;
}

.active_member {
  background-color: rgb(144, 144, 226);
  padding: 0.3rem;
}

.status_badge {
  background-color: rgb(97, 97, 97);
  color: white;
  padding: 0.3rem;
  font-size: 0.8rem;
}
