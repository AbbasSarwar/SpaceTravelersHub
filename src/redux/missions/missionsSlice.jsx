import { createSlice } from '@reduxjs/toolkit';
import { missionsData } from './MissionsAPI';

const initialState = {
  loading: false,
  missions: [],
  error: '',
};

export const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, { payload }) => {
     const missions = state.missions.map((mission) => {
        if (mission.id === payload) return { ...mission, reserved: true };
        return mission;
      });
      return { ...state, missions };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(missionsData.pending, (state) => {
        state.loading = true;
      })
      .addCase(missionsData.fulfilled, (state, {payload}) => ({
        ...state,
        loading: false,
        missions: payload,
      }))
      .addCase(missionsData.rejected, (state, {error}) => {
        state.loading = false;
        state.error = error.message;
      });
  },
});

export const { joinMission } = missionsSlice.actions;
export default missionsSlice.reducer;