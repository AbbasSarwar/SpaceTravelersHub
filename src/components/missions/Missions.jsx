import Navbar from '../Navbar';
import MissionsLogic from './MissionsLogic';
import '../../styles/missions/missions.css';

const Missions = () => (
  <>
    <Navbar />
    <MissionsLogic />
  </>
);

export default Missions;
