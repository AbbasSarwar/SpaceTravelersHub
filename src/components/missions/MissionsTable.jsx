import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { joinMission } from "../../redux/missions/missionsSlice";
import styles from "../../styles/Missions.module.css";

const MissionsTable = ({missions}) => {
  const dispatch = useDispatch();
  return ( 
    <table>
      <thead>
        <tr>
          <th className="mission">Misssion</th>
          <th className="description">Description</th>
          <th className="status">Status</th>
          <th className="join-leave-mission"></th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={mission.id}>
            <td>{mission.mission_name}</td>
            <td>{mission.description}</td>
            <td><span className="status-badge">NOT A MEMBER</span></td>
            <td><button className={`${!mission.reserved? styles.join: styles.leave}`} onClick={()=> dispatch(joinMission(mission.id))}>{!mission.reserved ? 'Join Mission': 'Leave Mission'}</button></td>
          </tr>
        ))}
      </tbody>
    </table>
   );
}

MissionsTable.propTypes = {
  missions: PropTypes.isRequired,
};

 
export default MissionsTable;