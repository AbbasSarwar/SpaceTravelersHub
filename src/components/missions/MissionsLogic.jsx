import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import MissionsTable from './MissionsTable';
import { missionsData } from '../../redux/missions/MissionsAPI';

const MissionsLogic = () => {
  const { loading, missions, error } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(missionsData());
  }, [dispatch]);

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>{error}</div>;
  }

  return (
    <div>
      <MissionsTable missions={missions} />
    </div>
  );
};

export default MissionsLogic;
