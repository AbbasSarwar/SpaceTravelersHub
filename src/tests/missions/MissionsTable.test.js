import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { MissionsTable } from '../../components/missions/MissionsTable';
import store from '../../redux/store';

it('renders correctly', () => {
  const missions = [
    { id: 1, mission_name: 'Mission 1', description: 'Description 1', reserved: true },
    { id: 2, mission_name: 'Mission 2', description: 'Description 2', reserved: false },
    // Add more mission objects as needed
  ];

  const { container } = render(
    <Provider store={store}>
      <BrowserRouter>
        <MissionsTable missions={missions} />
      </BrowserRouter>
    </Provider>
  );

  expect(container).toMatchSnapshot();
});
