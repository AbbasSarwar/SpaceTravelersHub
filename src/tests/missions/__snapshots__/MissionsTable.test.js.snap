// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          class="mission"
        >
          Misssion
        </th>
        <th
          class="description"
        >
          Description
        </th>
        <th
          class="status"
        >
          Status
        </th>
        <th
          class="join-leave-mission"
        />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Mission 1
        </td>
        <td>
          Description 1
        </td>
        <td>
          <span
            class="active_member"
          >
            Active Member
          </span>
        </td>
        <td>
          <button
            class="leave"
          >
            Leave Mission
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Mission 2
        </td>
        <td>
          Description 2
        </td>
        <td>
          <span
            class="status_badge"
          >
            NOT A MEMBER
          </span>
        </td>
        <td>
          <button
            class="join"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
